Rename oneMKL Interface to oneMath

Rbiessy opened the pull request:
# Description

Related RFC: https://github.com/oneapi-src/oneMKL/pull/564 and related spec PR: https://github.com/uxlfoundation/oneAPI-spec/pull/596
As a reminder the plan is to have this PR and the spec PR approved before we move this repository to https://github.com/uxlfoundation and merge the 2 PRs.

- The PR has the following implications on the MKLCPU and MKLGPU backends:
  - Had to introduce [`src/include/common_onemkl_conversion.hpp`](https://github.com/oneapi-src/oneMKL/pull/602/files#diff-1bd466e9a241e032faf9da384f46900bbc322e8258b086802c6f34d3a50ad381R36) to convert oneMath common types to oneMKL types. This wasn't needed before as the backends relied on the 2 projects to use the same namespace. The file also provide helper macros to catch oneMKL exceptions and rethrow them as oneMath exceptions.
  - The DFT MKLGPU backend used to catch and rethrow some but not all `oneapi::mkl` exceptions when committing a descriptor. We discussed this issue by email recently. With the changes needed by the renaming, we now catch and rethrow all the exceptions as `oneapi::math` exceptions and I was able to remove a related workaround in the tests (FYI @oneapi-src/onemkl-dft-write). Also see related discussion below: https://github.com/oneapi-src/oneMKL/pull/602#discussion_r1814235856
  - The RNG MKLGPU backend relied heavily on the 2 libraries using the same namespace. I introduced [`onemkl_distribution_conversion.hpp`](https://github.com/oneapi-src/oneMKL/pull/602/files#diff-6687d28034219fe472148c98659e2b98f391007028e2811f732b5823695cb33fR36) to convert the RNG types and call the backend's free functions such as `oneapi::mkl::rng::generate`, `oneapi::mkl::rng::skip_ahead`. @oneapi-src/onemkl-rng-write you may want to carefully review `src/rng/backends/mklgpu/` as I am not familiar with the RNG domain.
  - The renaming made it clear that some header files were duplicating function declarations for no reason in the BLAS and LAPACK domains. They are removed in https://github.com/oneapi-src/oneMKL/commit/baeb476f92b6dd7a0d137ab87c09fde1ec706634.
- The PR also remove the domain documentation for BLAS and LAPACK which duplicates the specification (commit https://github.com/oneapi-src/oneMKL/commit/b382291281b8a4b857d7f3e181fe7edf4862a252). We discussed in the issue https://github.com/oneapi-src/oneMKL/issues/489 that the duplicated documentation was an issue and some of us are keen to remove it. Now is a good time as it becomes even more outdated with the name change. (@oneapi-src/onemkl-blas-write and @oneapi-src/onemkl-lapack-write FYI)
- The changes assume that the teams `@oneapi-src/onemkl-*` will be renamed (or duplicated) to `@uxlfoundation/onemath-*` in the new repository (@rscohn2 FYI)
- The changes assume that the [`onemkl`](https://uxlfoundation.slack.com/archives/onemkl) Slack channel will be renamed to [`onemath`](https://uxlfoundation.slack.com/archives/onemath) (@rodburns FYI)
- Removed `USE_MKLREF` which was only used in Lapack tests and was not documented. This macro wouldn't work easily with the namespace change anymore (@oneapi-src/onemkl-lapack-write FYI)
- For reference the remaining occurrences of MKL are:
  - Intel backend names `MKLCPU` and `MKLGPU`.
  - Usages of `oneapi::mkl` inside the Intel backends.
  - File `cmake/mkl/MKLConfig.cmake` (could probably be removed IMO).
  - onemkl.maintainers@intel.com in CODE_OF_CONDUCT.md and CONTRIBUTING.md. Let me know if there is any plan to create onemath.maintainers@intel.com. If not this can be updated separately.
  - Folder `include/oneapi/mkl` for deprecated headers and deprecated `oneapi::mkl` C++ namespace.
  - Deprecated ONEMKL CMake namespace.
  - Some files in the `MKLCPU` and `MKLGPU` backends use "mkl" in their name. They could be updated to onemkl if needed.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally?
  - Local CPU and iGPU with MKLCPU and MKLGPU backends: [log_intel_cpu_igpu.txt](https://github.com/user-attachments/files/17463848/log_intel_cpu_igpu.txt)
  - Intel PVC: [log_pvc.txt](https://github.com/user-attachments/files/17463851/log_pvc.txt)
    - The BLAS and DFT tests failing are reproduced on the current develop branch, see GitHub issues https://github.com/oneapi-src/oneMKL/issues/600 and https://github.com/oneapi-src/oneMKL/issues/601
  - CUDA backends A100: [log_a100.txt](https://github.com/user-attachments/files/17463871/log_a100.txt)
    - The BLAS tests failing are reproduced on the develop branch, see GitHub issue https://github.com/oneapi-src/oneMKL/issues/599
  - AMD backends MI210: [log_mi210.txt](https://github.com/user-attachments/files/17463894/log_mi210.txt)
  - BLAS netlib: [log_netlib.txt](https://github.com/user-attachments/files/17463907/log_netlib.txt)
  - portBLAS and portFFT backends on Intel CPU: [log_portblas_portfft.txt](https://github.com/user-attachments/files/17463918/log_portblas_portfft.txt)
- [x] Have you formatted the code using clang-format? The PR already includes the clang-format changes from https://github.com/oneapi-src/oneMKL/pull/594 to avoid major conflicts when merging with develop.

Rbiessy committed:
Remove documentation duplicating the specification

Rbiessy committed:
Move and rename files

Rbiessy committed:
onemkl -> onemath

Rbiessy committed:
ONEMKL -> ONEMATH

Rbiessy committed:
Clarify comment BLAS -> CBLAS

Rbiessy committed:
oneMKL (Interfaces)? -> oneMath

Rbiessy committed:
Reword introduction

Rbiessy committed:
Clarify comments referring to MKL

Rbiessy committed:
Rename project CMake namespace MKL:: to ONEMATH::

Rbiessy committed:
Rename mkl namespace to math

Rbiessy committed:
Rename occurrences in third-party-programs/THIRD-PARTY-PROGRAMS

Rbiessy committed:
oneapi-src -> uxlfoundation

Rbiessy committed:
Rename oneMKL in docs/building_the_project_with_dpcpp.rst

Rbiessy committed:
Rename mkl in CONTRIBUTING.md

Rbiessy committed:
Rename oneMKL in READMEs

Rbiessy committed:
oneMath interface -> oneMath

Rbiessy committed:
Rename _MKL_ macro prefix to ONEMATH_

Rbiessy committed:
Add deprecated CMake targets

Rbiessy committed:
Add deprecated headers and mkl namespace

Rbiessy committed:
Rename oneMathConfig.cmake

Rbiessy committed:
Add include to Intel oneMKL lapack headers

Rbiessy committed:
Fix spblas MKL include

Rbiessy committed:
Fix compilation with MKL backends

Rbiessy committed:
Fix rng MKL backends

Rbiessy committed:
Rethrow oneapi::mkl exceptions as oneapi::math ones

Rbiessy committed:
Remove MKL_INCLUDE from curand and rocrand backends

Rbiessy committed:
fp_scalar_mkl -> fp_scalar_mkl

Rbiessy committed:
Rename mkl_*_table

Rbiessy committed:
Rename mkl after merge

Rbiessy committed:
Remove old workaround catch FFT_UNIMPLEMENTED

Rbiessy committed:
Factorize target_add_onemath_include

Rbiessy committed:
Use angle brackets to include Intel oneMKL headers

Rbiessy committed:
Update CODEOWNERS

Rbiessy committed:
Use oneAPI Math Library (oneMath)

Rbiessy committed:
clang-format-19.1

Rbiessy committed:
Remove USE_MKLREF

rafbiels commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -423,20 +422,20 @@ set, the backend libraries to enable the use of BLAS, LAPACK and RNG with MKLGPU
+portBLAS backend only, for a generic SYCL device supported by the Open DPC++
+project.
 
 Build oneMKL for the DFT domain on a generic SYCL device:
```
This DFT example section still needs renaming in the same way as the BLAS example above (three mentions of "oneMKL")

Rbiessy replied:
Thank you for catching them! This is fixed in https://github.com/oneapi-src/oneMKL/pull/602/commits/a42a97089dae9a1e672a1560b0bc4545cea64cd9

rafbiels commented on docs/README.md:
This file still says `<path to onemkl>` twice (meaning this project)

Rbiessy replied:
Fixed in https://github.com/oneapi-src/oneMKL/pull/602/commits/a42a97089dae9a1e672a1560b0bc4545cea64cd9

rafbiels commented on docs/building_the_project_with_adaptivecpp.rst:
```diff
@@ -35,7 +35,7 @@ On Linux (other OSes are not supported with the AdaptiveCpp compiler):
 
   # Inside <path to onemkl>
```
This file still says `<path to onemkl>` (meaning this project) here and in the "Environment Setup" above

Rbiessy replied:
Fixed in https://github.com/oneapi-src/oneMKL/pull/602/commits/a42a97089dae9a1e672a1560b0bc4545cea64cd9

rafbiels commented on examples/README.md:
I'm not sure why we have this as an example, but in the DFT portFFT example output in this file, it says:
```
	oneMKL: dft/backends/portfft/commit: function is not implemented portFFT only supports complex to complex transforms
```
so if we keep this example (showing this doesn't work) then we should update the message here.

Rbiessy replied:
Yeah I know this is not the best example. The issue is that portFFT is experimental and we use the same examples for all the domains. I think it's fine.
I re-ran the example and updated the whole message in https://github.com/oneapi-src/oneMKL/pull/602/commits/a42a97089dae9a1e672a1560b0bc4545cea64cd9

rafbiels reviewed (commented):
That's an immense amount of work, well done! I only found a few missed occurrences.

Rbiessy committed:
Merge branch 'develop' into romain/rename

Rbiessy committed:
Rename missing oneMKL occurrences

Rbiessy commented:
I merged the PR with the develop branch and the clang-format-19.10 changes which reduces the diff.
I have also updated the PR description with more details on the implications to make it work with the MKLCPU and MKLGPU backends (first bullet point).

anantsrivastava30 commented on src/dft/backends/mklgpu/commit.cpp:
```diff
@@ -23,118 +23,98 @@
-            }
-            catch (const std::exception& mkl_exception) {
-                // Catching the real Intel oneMKL exception causes headaches with naming.
-                if (!forward_good) {
```
How will the removal of the forward_good impact the logic incase the one_descriptor is false, (we still support in/output_strides in deprication)?

Rbiessy replied:
That's a good question. The point of `forward_good` was to make sure we don't throw an exception when at least one direction is valid. I know we had some tests that used to need this as the Intel backend used to throw exceptions as one configuration was invalid for the particular direction and not the other. I have noticed this is not needed anymore, probably because we use the parameters `FWD_STRIDES` and `BWD_STRIDES` now but I haven't dug deeper than this.
The renaming was a good opportunity to simplify this as I need to rethrow the exceptions as their `oneapi::math` equivalent. Let me know if you think this change is a risk.

anantsrivastava30 replied:
If we are only using the `fwd/bwd_strides` then its fine to clean up this logic, I don't think this change is a risk as we want to use the strides config anyway and move away from the `input/output_distance`. Perhaps it would be wise to run the DFT GPU tests for this PR as well (as I only see only the CPU unit tests have run in post-commit.

Rbiessy replied:
Ok, good. The logs are attached in the PR's description. One of them is running on PVC. There are some failures that can also be reproduced on the develop branch.

anantsrivastava30 commented on include/oneapi/math/detail/backends_table.hpp:
```diff
@@ -138,78 +138,78 @@ static std::map<domain, std::map<device, std::vector<const char*>>> libraries =
-                                                     { domain::lapack, "mkl_lapack_table" },
-                                                     { domain::dft, "mkl_dft_table" },
-                                                     { domain::rng, "mkl_rng_table" },
+static std::map<domain, const char*> table_names = { { domain::blas, "onemath_blas_table" },
```
why onemath_<domain> instead of math_<domain>? "one" is slightly intel releated can't we just have math_<domain>? Same comment in the device_wrappers.cpp: [dft_function_table_t](https://github.com/oneapi-src/oneMKL/blob/1462c0b72f5599f2cdf184ae8a0a100f621f554d/src/dft/backends/mklcpu/mkl_dft_cpu_wrappers.cpp#L26)

Rbiessy replied:
These occurrences really should have been renamed to `onemkl` earlier as they mean to refer to this project. Similarly I much prefer to use `onemath` instead of `math` when possible to clarify we refer to this project and not "mathematics". These variables aren't used often so 3 more characters should be fine. I'll consider changing it if more people ask for it.

anantsrivastava30 reviewed (commented):
These are a lot of changes for one PR!. Mostly they look fine with what was discussed in the RFC and spce change. However I have a few question in my review.

Rbiessy commented:
The conflicts are mostly due to the recent PR for cuSPARSE support. I will work on fixing them but the other domains can still be reviewed.

andreyfe1 commented:
Thanks for doing this huge work!
I have a general question from my side. Have you used `git mv` for renaming/duplicating files? This command preserves the file’s history. So, we will be able to see the history of commits for old files.

Rbiessy commented:
> Thanks for doing this huge work! I have a general question from my side. Have you used `git mv` for renaming/duplicating files? This command preserves the file’s history. So, we will be able to see the history of commits for old files.

Yes I have used `git mv` in the second commit: https://github.com/oneapi-src/oneMKL/pull/602/commits/68d3966da511857c4b4a2eb8526510a822db4d53
In this commit GitHub shows the movement unfortunately when the changes from the other commits are shown GitHub decides it's best to show some files have been deleted and re-created.
To work around this you could try to compare the diff between the commit after the move and the latest commit if that's easier. You can see it here: https://github.com/Rbiessy/oneMKL/compare/600cd44...2a8a76b?diff=split&w=

andreyfe1 commented:
ok. Thanks. Just wanted to make sure that we don't accidentally miss important information

Rbiessy committed:
Merge branch 'develop' into romain/rename

Rbiessy commented:
@oneapi-src/onemkl-sparse-write I have updated the PR to fix the conflicts with the cuSPARSE PR. It would be useful if you can review the sparse related changes in this PR before the rocSPARSE PR https://github.com/oneapi-src/oneMKL/pull/544!
Log for Intel backends: [intel_mklcpu_mklgpu.txt](https://github.com/user-attachments/files/17636353/intel_mklcpu_mklgpu.txt)
Log for Nvidia backends: [log_a100.txt](https://github.com/user-attachments/files/17636357/log_a100.txt)

Rbiessy committed:
Merge branch 'develop' into romain/rename

Rbiessy committed:
Remove duplicated blas and lapack headers

Rbiessy commented:
I fixed some more recent conflicts.
@oneapi-src/onemkl-blas-write, @oneapi-src/onemkl-lapack-write in the last commit (https://github.com/oneapi-src/oneMKL/pull/602/commits/baeb476f92b6dd7a0d137ab87c09fde1ec706634) I have removed a couple of header files in blas and lapack which were duplicating function declarations for the MKLCPU and MKLGPU backends. Given that we can simply include the Intel oneMKL headers instead they can be removed here. This was discussed in https://github.com/oneapi-src/oneMKL/pull/606#issuecomment-2446361272
Log testing the MKLCPU and MKLGPU backends with the 2025.0 base toolkit: [intel_mklcpu_mklgpu.txt](https://github.com/user-attachments/files/17663292/intel_mklcpu_mklgpu.txt)

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -3,32 +3,31 @@
-oneAPI DPC++ Compiler or open-source oneAPI DPC++ Compiler. For guidance on
-building the project with AdaptiveCpp, see
-:ref:`building_the_project_with_adaptivecpp`.
+This page describes building the oneMath with either the Intel(R) oneAPI DPC++
```
```suggestion
This page describes building oneMath with either the Intel(R) oneAPI DPC++
```

Rbiessy replied:
Thanks, that's fixed in https://github.com/oneapi-src/oneMKL/pull/602/commits/463fe574ce1e90d3314078b12551f0529c637167
It doesn't look like there were any other occurrences like this.

andrewtbarker reviewed (approved):
BLAS changes look good to me.

sknepper commented:
/intelci: run

Rbiessy committed:
the oneMath -> oneMath

anantsrivastava30 reviewed (approved):
Approving on behalf of DFT

toxicscum commented:
/intelci: run

sknepper commented:
Thanks for all your work here - there are so many changes, so I know you've spent a lot of effort on this!
It looks like there are some problems building on Windows. For example, I'm seeing an error for the BLAS domain like:

[1/155] C:\cache\stash\oneapi-compiler\20240719_rls\win\package\compiler\latest\bin\icx.exe -fsycl /nologo  -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src\include -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\build\bin -IC:\cache\stash\onemkl\regular\2024.2.1\20240722\build\__release_win\include /EHsc /DWIN32 /D_WINDOWS /W3 /GR /EHsc -Wno-unused-function -w  /O2 /Ob2 /DNDEBUG -iquote C:/temp/ec/mkltest/auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d/sources/include -DMKL_ILP64 -fsycl -QMD -QMT bin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj -QMF bin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj.d /Fobin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj -c C:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src\blas\backends\mklcpu\mklcpu_level2.cpp
FAILED: bin/blas/backends/mklcpu/CMakeFiles/onemath_blas_mklcpu_obj.dir/mklcpu_level2.cpp.obj
C:\cache\stash\oneapi-compiler\20240719_rls\win\package\compiler\latest\bin\icx.exe -fsycl /nologo  -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src\include -IC:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\build\bin -IC:\cache\stash\onemkl\regular\2024.2.1\20240722\build\__release_win\include /EHsc /DWIN32 /D_WINDOWS /W3 /GR /EHsc -Wno-unused-function -w  /O2 /Ob2 /DNDEBUG -iquote C:/temp/ec/mkltest/auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d/sources/include -DMKL_ILP64 -fsycl -QMD -QMT bin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj -QMF bin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj.d /Fobin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj -c C:\temp\ec\mkltest\auto-efa034fe-1902-f165-9fcf-a4bf010d0e2d\sources\src\blas\backends\mklcpu\mklcpu_level2.cpp
icx: error: cannot specify '-Fobin\blas\backends\mklcpu\CMakeFiles\onemath_blas_mklcpu_obj.dir\mklcpu_level2.cpp.obj' when compiling multiple source files

And somewhat similar for LAPACK (and RNG); e.g.,
icx: error: cannot specify '-Fobin\lapack\backends\mklcpu\CMakeFiles\onemath_lapack_mklcpu_obj.dir\mkl_lapack.cpp.obj' when compiling multiple source files

Rbiessy committed:
Fix compilation on Windows

Rbiessy commented:
Thanks @sknepper I have been able to identify that the issue is due to using `-iquote` flag [here](https://github.com/oneapi-src/oneMKL/pull/602/files#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R57) which is not supported on Windows.
This is needed because we now need to include the Intel oneMKL `mkl.hpp` file to have a declaration of the `oneapi::mkl` symbols. Including this file conflicts with the `mkl.hpp` in this project which does not provide the right symbols anymore.
I am looking for another solution other than using `-iquote` but I am not confident I can find a nice solution. This is a difficult problem to solve and the main reason why I think different projects should not provide the same header files.

In terms of review it is still possible to review most of the changes. I expect the changes needed will only affect CMake files.

Rbiessy commented:
@toxicscum @sknepper I have found a decent solution which should fix the compilation issue on Windows in https://github.com/oneapi-src/oneMKL/pull/602/commits/a2de0e5517ba2bbec30d0efe690130997efcf2c6.
I have described the issue and the solution I used in this comment: https://github.com/oneapi-src/oneMKL/pull/602/commits/a2de0e5517ba2bbec30d0efe690130997efcf2c6#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R51

sknepper commented:
/intelci: run

sknepper reviewed (approved):
Thanks, @Rbiessy - what an impressive set of changes!
LAPACK looks good - thanks for removing the duplicated headers and documentation - that will streamline things in the future.
The USE_MKLREF in the tests was just a convenience feature to avoid a dependency on Netlib; in the future, if we create standalone testing, the same goal will have been achieved. So that's fine to remove this undocumented macro.

The latest round of precommit testing is still running, but it does indeed look like the Windows build issues were resolved - much thanks!

Rbiessy committed:
Merge branch 'develop' into romain/rename

mkrainiuk commented on README.md:
```diff
@@ -1,25 +1,25 @@
+# oneAPI Math Library (oneMath)
 
-oneMKL Interfaces is an open-source implementation of the oneMKL Data Parallel C++ (DPC++) interface according to the [oneMKL specification](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemkl/source/). It works with multiple devices (backends) using device-specific libraries underneath.
+oneMath is an open-source implementation of the [oneMath specification](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemath/source/). It can work with multiple devices using multiple libraries (backends) underneath. The oneMath project was previously referred to as oneMKL Interface.
```
Small improvement:
```suggestion
oneMath is an open-source implementation of the [oneMath specification](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemath/source/). It can work with multiple devices using multiple libraries (backends) underneath. The oneMath project was previously referred to as oneMKL Interfaces.
```

mkrainiuk commented on README.md:
```diff
@@ -544,30 +544,30 @@ Product | Supported Version | License
 ## Contributing
 
-You can contribute to this project and also contribute to [the specification for this project](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemkl/source/). Please read the [CONTRIBUTING](CONTRIBUTING.md) page for more information. You can also contact oneMKL developers and maintainers via [UXL Foundation Slack](https://slack-invite.uxlfoundation.org/) using [#onemk ...
+You can contribute to this project and also contribute to [the specification for this project](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onemath/source/). Please read the [CONTRIBUTING](CONTRIBUTING.md) page for more information. You can also contact oneMath developers and maintainers via [UXL Foundation Slack](https://slack-invite.uxlfoundation.org/) using [#one ...
```
Do you know who can update the slack channel name? Does it make sense to introduce new channel and close the old one once these changes will be merged?

rodburns replied:
> Do you know who can update the slack channel name? Does it make sense to introduce new channel and close the old one once these changes will be merged?

I can rename the channel or create a new one. It might make sense to retain the existing channel though I am open to whatever the consensus is. Just drop me a message when you are ready.

mkrainiuk replied:
Thanks @rodburns! Renaming'd be perfect option if possible.

mkrainiuk reviewed (approved):
Changes look good to me, thank you! I have two minor comments.

Rbiessy committed:
Interface -> Interfaces

Co-authored-by: Maria Kraynyuk <maria.kraynyuk@intel.com>

Rbiessy commented:
@spencerpatty gentle ping since you have approved the PR for the specification renaming but not the implementation.
Also pinging @oneapi-src/onemkl-rng-write if you are able to review the 2 renaming PRs before November 29 that'd be great.
Thanks

spencerpatty commented on src/sparse_blas/backends/mkl_common/mkl_spmv.cxx:
```diff
@@ -191,34 +200,36 @@ sycl::event internal_spmv(sycl::queue& queue, oneapi::mkl::transpose opA, const
-            return oneapi::mkl::sparse::trmv(queue, A_view.uplo_view, opA, A_view.diag_view,
-                                             host_alpha, backend_handle, x_usm, host_beta, y_usm,
-                                             dependencies);
+            RETHROW_ONEMKL_EXCEPTIONS_RET(
```
nice macro for rethrowing exceptions and returning sycl events :)  handy that!

Rbiessy replied:
Well I wish this could have been a templated function but it was too tricky to get it working with functions that return an event and functions that don't unfortunately. Glad you like it though!

spencerpatty commented on src/sparse_blas/backends/mkl_common/mkl_spsv.cxx:
```diff
@@ -45,16 +43,16 @@ sycl::event release_spsv_descr(sycl::queue& queue, spsv_descr_t spsv_descr,
 
-void check_valid_spsv(const std::string& function_name, oneapi::mkl::transpose opA,
-                      matrix_view A_view, matrix_handle_t A_handle, dense_vector_handle_t x_handle,
+void check_valid_spsv(const std::string& function_name, transpose opA, matrix_view A_view,
```
much simpler now to distinguish between oneMKL product types and the oneMATH types :)  good changes for this interfaces project

spencerpatty reviewed (approved):
I have reviewed all Sparse BLAS related files and I **approve** of these changes.  I love that it is much simpler to distinguish whether we are using oneMATH variables or the backend variables in the files now.  This is a good thing!

spencerpatty commented:
> @oneapi-src/onemkl-sparse-write I have updated the PR to fix the conflicts with the cuSPARSE PR. It would be useful if you can review the sparse related changes in this PR before the rocSPARSE PR #544! Log for Intel backends: [intel_mklcpu_mklgpu.txt](https://github.com/user-attachments/files/17636353/intel_mklcpu_mklgpu.txt) Log for Nvidia backends: [log_a100.txt](https://github.com/user-attachments/files/17636357/log_a100.txt)

@Rbiessy I have reviewed  the cuSPARSE changes and the general changes for Sparse BLAS and approved them, but in the intel mklcpu/gpu logs it seems like all the SPARSE BLAS tests are skipping.  Is that intended?  cuSPARSE tests appear to run the first 4 or 5 then skip the rest as well ...  is that intended ?

Rbiessy commented:
> @Rbiessy I have reviewed the cuSPARSE changes and the general changes for Sparse BLAS and approved them, but in the intel mklcpu/gpu logs it seems like all the SPARSE BLAS tests are skipping. Is that intended? cuSPARSE tests appear to run the first 4 or 5 then skip the rest as well ... is that intended ?

@spencerpatty yes, this is expected. Unfortunately the tests are not "unit test" so each GTest is actually running a dozen or more smaller tests. See for instance the spmm [tests](https://github.com/oneapi-src/oneMKL/blob/develop/tests/unit_tests/sparse_blas/include/test_spmm.hpp#L89), each call to `test_functor_i32` is testing a different configuration of spmm. If any of these configurations is not supported the GTest will be marked as skipped. I think it would be better if each GTest tests one configuration but:

1. It would differ from most other domains.
2. We would need some more work to re-use the SYCL queue across GTest to avoid the expensive creation of the queue. This is possible but it requires to use test fixtures with `TEST_F` which I believes makes it impossible to easily generate the tests for different data types.

We experimented with the smaller unit tests approach in the DFT domain but didn't find a nice solution to re-use the SYCL queue. This becomes an issue the more tests we have.

We made sure that the configurations that are skipped are expected to be skipped when reviewing the PRs for the Intel backends.

Rbiessy committed:
Merge branch 'develop' into romain/rename

Rbiessy commented:
I fixed some minor conflicts with the rng domain, see the rng tests log attached: [log_rng.txt](https://github.com/user-attachments/files/17992711/log_rng.txt)

Thanks all for the feedback, we're proceeding with the migration today!
